File: c#/src/Sodium/Sodium.Tests/PerformanceTests.cs
using System;
using System.Linq;
using NUnit.Framework;

namespace Sodium.Tests
{
    [TestFixture]
    public class PerformanceTests
    {
        [Test]
        public void TestMerge()
        {
            StreamSink<Unit> s = Stream.CreateSink<Unit>();
            var obj = Transaction.Run(() =>
            {
                StreamLoop<bool> loop = Stream.CreateLoop<bool>();
                DiscreteCellStreamSink<int> s1 = new DiscreteCellStreamSink<int>();
                DiscreteCellStreamSink<int> s2 = new DiscreteCellStreamSink<int>();
                TestObject[] l = Enumerable.Range(0, 5000).Select(_ => new TestObject(loop, s1, s2)).ToArray();
                loop.Loop(s.Snapshot(l.Select(o => o.Cell).Lift().Cell).Map(o => o.All(v => v == 0)));
                return l;
            });
            int[] values = obj.Select(v => v.CurrentValue).ToArray();
        }

        private class TestObject
        {
            // ReSharper disable once NotAccessedField.Local
            private readonly IListener l;
            private Lazy<int> currentValue = new Lazy<int>(() => default(int));

            public TestObject(Stream<bool> s, Stream<int> s1, Stream<int> s2)
            {
                (DiscreteCell<int> cell, IStrongListener l) = Transaction.Run(() =>
                {
                    var cellLocal = s.Map(v => v ? 1 : 0).OrElse(s1).OrElse(s2).Hold(0);
                    var cell2 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell3 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell4 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell5 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell6 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell7 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell8 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);
                    var cell9 = s1.Snapshot(cellLocal.Cell, (left, right) => left + right).Filter(v => v > 5).OrElse(s.Snapshot(s1.Hold(0).Lift(s2.Hold(1), (left, right) => left + right).Cell).Map(v => v + 1)).Hold(3);

                    var lLocal = Transaction.Run(() =>
                    {
                        this.currentValue = cellLocal.Cell.SampleLazy();
                        return cellLocal.Updates.Listen(v => this.CurrentValue = v);
                    });

                    return (cellLocal, lLocal);
                });

                this.Cell = cell;
                this.l = l;
            }

            public DiscreteCell<int> Cell { get; }

            public int CurrentValue
            {
                get { return this.currentValue.Value; }
                private set { this.currentValue = new Lazy<int>(() => value); }
            }
        }
    }
}

File: f#/src/SodiumFRP.FSharp/Tests/PerformanceTests.fs
module SodiumFRP.FSharp.Tests.Performance

open System.Collections.Generic
open NUnit.Framework
open SodiumFRP.FSharp
open SodiumFRP.FSharp.Sodium

[<AutoOpen>]
module private Types =
    
    type TestObject (s : Stream<bool>, s1 : Stream<int>, s2 : Stream<int>) as this =
        let mutable currentValue = lazy 0
        let (cell, l) = runT (fun () ->
            let cell = ((s |> mapS (fun v -> if v then 1 else 0), s1) |> orElseS, s2) |> orElseS |> holdS 0
            let createCell () =
                (
                    s1 |> snapshotC cell (+) |> filterS (flip (>) 5),
                    s |> snapshotAndTakeC ((s1 |> holdS 0, s2 |> holdS 1) |> lift2C (+)) |> mapS ((+) 1)
                ) |> orElseS |> holdS 3
            let cell2 = createCell ()
            let cell3 = createCell ()
            let cell4 = createCell ()
            let cell5 = createCell ()
            let cell6 = createCell ()
            let cell7 = createCell ()
            let cell8 = createCell ()
            let cell9 = createCell ()
            currentValue <- cell |> sampleLazyC
            let l = cell |> updatesC |> listenS (fun v -> this.CurrentValue <- v)
            (cell, l))
        member private __.L = l
        member __.Cell = cell
        member __.CurrentValue with get () = currentValue.Value and set value = currentValue <- lazy value

[<TestFixture>]
type Tests() =

    [<Test>]
    member __.``Test Merge``() =
        let s = sinkS<unit> ()
        let struct (_, obj) = loopS (fun loop ->
            let s1 = sinkCS ()
            let s2 = sinkCS ()
            let l = Array.init 250 (fun _ -> TestObject (loop, s1, s2))
            struct (s |> snapshotAndTakeC (l |> Seq.map (fun o -> o.Cell) |> liftAllC id) |> mapS (Seq.forall (fun v -> v = 0)), l))
        let values = obj |> Array.map (fun o -> o.CurrentValue)
        CollectionAssert.AreEqual (Seq.init 250 (fun _ -> 0), values)